airbyte-integrations/connectors/source-zuora/source_zuora/source.py

```diff
@@ -0,0 +1,166 @@
+        """
+        return stream_state.get(self.cursor_field, stream_state.get(self.alt_cursor_field)) if stream_state else self.api.start_date
+
+    def read_records(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping]:
+        # if stream_state is missing, we will use the start-date from config for a full refresh
+        stream_state = self._get_stream_state(stream_state)
+        try:
+            yield from self.api._get_data_with_date_slice("select", self.name, self.cursor_field, stream_state, self.window_in_days)
+        except ZOQLQueryFieldCannotResolve:
+            self.cursor_field = self.alt_cursor_field
+            yield from self.api._get_data_with_date_slice("select", self.name, self.cursor_field, stream_state, self.window_in_days)
+        except ZOQLQueryCannotProcessObject:
+            pass
+
+
+# Incremental-refresh for Zuora Stream handels both full-refresh and incremental-refresh
+class IncrementalZuoraStream(ZuoraStream):
+
+    # Define default cursor field
+    cursor_field = "updateddate"
+    alt_cursor_field = "createddate"
+
+    # setting limit of the date-slice for the data query job
+    @property
+    def window_in_days(self) -> int:
+        return self.api.window_in_days
+
+    # setting checkpoint interval to the limit of date-slice
+    state_checkpoint_interval = window_in_days
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        return {self.cursor_field: max(latest_record.get(self.cursor_field, ""), current_stream_state.get(self.cursor_field, ""))}
+
+
+# Basic Connections Check
+class SourceZuora(AbstractSource):
+    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, any]:
+        """
+        Testing connection availability for the connector by granting the token.
+        """
+        auth = ZuoraAuthenticator(config["is_sandbox"]).generateToken(config["client_id"], config["client_secret"])
+        if auth.get("status") == 200:
+            return True, None
+        else:
+            return False, auth["status"]
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        """
+        Mapping a input config of the user input configuration as defined in the connector spec.
+        Defining streams to run.
+        """
+
+        def create_stream_class_from_object_name(zuora_objects: List) -> List:
+            """
+            The function to produce the dynamic stream classes from the list of zuora objects names
+            """
+            # Define the bases
+            cls_base = (IncrementalZuoraStream,)
+            cls_props = {}
+            # Build the streams
+            streams = []
+            for obj in zuora_objects:
+                # List of zuora objects >> stream classes
+                streams.append(type(obj, cls_base, cls_props))
```

Comment:
Thanks, this approach looks reasonable to use here, because we have many similar classes by their structure, a lot of potential streams, custom objects and we need to define them dynamically based on the account setting of the user.